src/libAtomVM/ets.c

```diff
@@ -252,58 +252,102 @@ static void ets_delete_all_tables(struct Ets *ets, GlobalContext *global)
     ets_delete_tables_internal(ets, true_pred, NULL, global);
 }
 
-EtsErrorCode ets_insert(term ref, term entry, Context *ctx)
+static EtsErrorCode ets_table_insert(struct EtsTable *ets_table, term entry, Context *ctx)
 {
-    struct EtsTable *ets_table = term_is_atom(ref) ? ets_get_table_by_name(&ctx->global->ets, ref, TableAccessWrite) : ets_get_table_by_ref(&ctx->global->ets, term_to_ref_ticks(ref), TableAccessWrite);
-    if (ets_table == NULL) {
-        return EtsTableNotFound;
-    }
-
     if (ets_table->access_type != EtsAccessPublic && ets_table->owner_process_id != ctx->process_id) {
-        SMP_UNLOCK(ets_table);
         return EtsPermissionDenied;
     }
 
     if ((size_t) term_get_tuple_arity(entry) < (ets_table->keypos + 1)) {
-        SMP_UNLOCK(ets_table);
         return EtsBadEntry;
     }
 
-    Heap *heap = malloc(sizeof(Heap));
-    if (IS_NULL_PTR(heap)) {
-        SMP_UNLOCK(ets_table);
+    int keypos = (int) ets_table->keypos;
+
+    struct HNode *new_node = ets_hashtable_new_node(entry, keypos, ctx->global);
+    if (IS_NULL_PTR(new_node)) {
         return EtsAllocationFailure;
     }
-    size_t size = (size_t) memory_estimate_usage(entry);
-    if (memory_init_heap(heap, size) != MEMORY_GC_OK) {
-        free(heap);
-        SMP_UNLOCK(ets_table);
-        return EtsAllocationFailure;
+
+    EtsErrorCode result = EtsOk;
+    EtsHashtableErrorCode res = ets_hashtable_insert(ets_table->hashtable, new_node, EtsHashtableAllowOverwrite, ctx->global);
+    if (UNLIKELY(res != EtsHashtableOk)) {
+        result = EtsAllocationFailure;
```

Comment:
Just `return EtsAllocationFailure` and `return EtsOk` below, no need for additional variable.